src/provider/interface.ts

```diff
@@ -330,4 +332,18 @@ export abstract class ProviderInterface {
    * @returns StateUpdateResponse
    */
   public abstract getStateUpdate(blockIdentifier?: BlockIdentifier): Promise<StateUpdateResponse>;
+
+  /**
+   * Gets the contract version from the provided address
+   * @param contractAddress required if no classHash
+   * @param classHash required if no contractAddress
+   * @param options - getContractVersionOptions
+   *   - (optional) compiler - (default true) extract compiler version using type tactic from abi
+   *   - (optional) blockIdentifier - block identifier
+   */
+  public abstract getContractVersion(
+    contractAddress?: string,
+    classHash?: string,
+    options?: getContractVersionOptions
+  ): Promise<ContractVersion>;
```

Comment:
If I do overload  (I think TS implementation of it is awful), then only full overload makes sense. And after all this addition declarations I still can do:
```ts
testProvider.getContractVersion(undefined, undefined);
```
And TS do not complain I need to have one-of

Reply:
If the changes from this comment suggestion are applied and the `testProvider` is cast as a `ProviderInterface` with `const testProvider = new Provider(getTestProvider()) as ProviderInterface` it should complain:
<img width="699" alt="Screenshot 2023-09-26 at 09 21 59" src="https://github.com/starknet-io/starknet.js/assets/8030344/9b2048de-22eb-4a79-82c7-66cc325dfa67">

For it to apply without the cast the overload it would need to be added as a prefix to the `Provider` implementation (with the mentioned `any` cast in the implementation body for passing the parameters):
```
  public async getContractVersion(
    contractAddress: string,
    classHash?: undefined,
    options?: getContractVersionOptions
  ): Promise<ContractVersion>;
  public async getContractVersion(
    contractAddress: undefined,
    classHash: string,
    options?: getContractVersionOptions
  ): Promise<ContractVersion>;
  public async getContractVersion(
    contractAddress?: string,
    classHash?: string,
    options?: getContractVersionOptions
  ) {
    return this.provider.getContractVersion(contractAddress as any, classHash as any, options);
  }
```
Same thing for `RpcProvider` and `SequencerProvider`, however, those wouldn't need changes in the implementation body.

So the parameter relationship can be enforced explicitly through typing but comes with the trade-off of increased verbosity (and prettier makes it more pronounced). I'll approve the PR and you can decide if the trade-off is worth it.